File: src/lib/threads-api.ts
const THREADS_API_BASE = 'https://graph.threads.net/v1.0';

export interface ThreadsApiConfig { accessToken: string; userId: string; proxyUrl?: string; }
export interface PublishResult { success: boolean; threadId?: string; error?: string; }

export async function exchangeCodeForToken(code: string, appId: string, appSecret: string, redirectUri: string): Promise<{ accessToken: string; userId: string }> {
  const response = await fetch(`${THREADS_API_BASE}/oauth/access_token`, { method: 'POST', headers: { 'Content-Type': 'application/x-www-form-urlencoded' }, body: new URLSearchParams({ client_id: appId, client_secret: appSecret, grant_type: 'authorization_code', redirect_uri: redirectUri, code }) });
  const data = await response.json();
  if (data.error) throw new Error(data.error.message);
  return { accessToken: data.access_token, userId: data.user_id };
}

export async function publishThread(config: ThreadsApiConfig, text: string, options?: { imageUrl?: string; videoUrl?: string; replyToId?: string; }): Promise<PublishResult> {
  try {
    const body: Record<string, string> = { media_type: 'TEXT', text, access_token: config.accessToken };
    if (options?.replyToId) body.reply_to_id = options.replyToId;
    const r1 = await fetch(`${THREADS_API_BASE}/${config.userId}/threads`, { method: 'POST', headers: { 'Content-Type': 'application/x-www-form-urlencoded' }, body: new URLSearchParams(body) });
    const d1 = await r1.json();
    if (d1.error) throw new Error(d1.error.message);
    const containerId = d1.id;
    await new Promise(resolve => setTimeout(resolve, 5000));
    const r2 = await fetch(`${THREADS_API_BASE}/${config.userId}/threads_publish`, { method: 'POST', headers: { 'Content-Type': 'application/x-www-form-urlencoded' }, body: new URLSearchParams({ creation_id: containerId, access_token: config.accessToken }) });
    const d2 = await r2.json();
    if (d2.error) throw new Error(d2.error.message);
    return { success: true, threadId: d2.id };
  } catch (error) {
    return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
  }
}

export function processSpintax(text: string): string {
  const regex = /\{([^{}]+)\}/g;
  return text.replace(regex, (_, options: string) => { const choices = options.split('|'); return choices[Math.floor(Math.random() * choices.length)]; });
}

export async function getUserProfile(config: ThreadsApiConfig) {
  const params = new URLSearchParams({ fields: 'id,username,threads_profile_picture_url,threads_biography', access_token: config.accessToken });
  const r = await fetch(`${THREADS_API_BASE}/${config.userId}?${params}`);
  const data = await r.json();
  if (data.error) throw new Error(data.error.message);
  return data;
}

export async function getPublishingLimit(config: ThreadsApiConfig) {
  const params = new URLSearchParams({ fields: 'quota_usage,config', access_token: config.accessToken });
  const r = await fetch(`${THREADS_API_BASE}/${config.userId}/threads_publishing_limit?${params}`);
  const data = await r.json();
  if (data.error) throw new Error(data.error.message);
  return data;
}

File: threads_api.py
"""
Threads API Module
"""
import aiohttp
import asyncio
import random
import re

API = "https://graph.threads.net/v1.0"


def process_spintax(text: str) -> str:
    pattern = r'\{([^{}]+)\}'
    def repl(m):
        return random.choice(m.group(1).split('|'))
    result = text
    while re.search(pattern, result):
        result = re.sub(pattern, repl, result)
    return result


async def publish_thread(user_id, token, text, reply_to_id=None, image_url=None):
    try:
        async with aiohttp.ClientSession() as s:
            payload = {
                "media_type": "IMAGE" if image_url else "TEXT",
                "text": text,
                "access_token": token
            }
            if image_url:
                payload["image_url"] = image_url
            if reply_to_id:
                payload["reply_to_id"] = reply_to_id

            async with s.post(f"{API}/{user_id}/threads", data=payload) as r:
                d = await r.json()
                if "error" in d:
                    return {"success": False, "error": d["error"].get("message", "")}
                cid = d["id"]

            await asyncio.sleep(5 if not image_url else 30)

            async with s.post(f"{API}/{user_id}/threads_publish", data={"creation_id": cid, "access_token": token}) as r:
                d = await r.json()
                if "error" in d:
                    return {"success": False, "error": d["error"].get("message", "")}
                return {"success": True, "thread_id": d["id"]}
    except Exception as e:
        return {"success": False, "error": str(e)}


async def get_user_profile(user_id, token):
    async with aiohttp.ClientSession() as s:
        async with s.get(f"{API}/{user_id}", params={"fields": "id,username", "access_token": token}) as r:
            return await r.json()


async def get_long_lived_token(short_token, app_secret):
    async with aiohttp.ClientSession() as s:
        async with s.get(
            "https://graph.threads.net/access_token",
            params={"grant_type": "th_exchange_token", "client_secret": app_secret, "access_token": short_token}
        ) as r:
            return await r.json()


async def refresh_token(token):
    async with aiohttp.ClientSession() as s:
        async with s.get(
            "https://graph.threads.net/refresh_access_token",
            params={"grant_type": "th_refresh_token", "access_token": token}
        ) as r:
            return await r.json()


async def get_publishing_limit(user_id, token):
    async with aiohttp.ClientSession() as s:
        async with s.get(
            f"{API}/{user_id}/threads_publishing_limit",
            params={"fields": "quota_usage,config", "access_token": token}
        ) as r:
            return await r.json()
